### PR TITLE
A large refactor

### PR DESCRIPTION
**Pull request by nerd0**
This is [PR #7](https://github.com/m3db/m3db-operator/pull/7) but on a new branch so that CI pick it up.

### CODE REVIEW

File: pkg/k8sops/generators_test.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+			Selector:  k.GenerateMaps("selectors", svcCfg),
+			Ports:     k.GenerateServicePorts(svcCfg.Ports),
+			ClusterIP: "None",
+		},
+	}
+	newSvc := k.GenerateService(svcCfg)
+	require.Equal(t, svc, newSvc)
+}
+
+func TestGenerateStatefulSet(t *testing.T) {
```
Comment (nerd0) [inline pkg/k8sops/generators_test.go]: Brittle test and function.

File: pkg/k8sops/generators.go [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+	_configurationName      = "m3-configuration"
+	_configurationFileName  = "m3.yml"
+)
+
+var (
+	_configurationFileLocation = fmt.Sprintf("%s%s", _configurationDirectory, _configurationFileName)
+)
+
+// GenerateCRD generates the crd object needed for the M3DBCluster
+func (k *k8sops) GenerateCRD() *apiextensionsv1beta1.CustomResourceDefinition {
```
Comment (schallert) [inline pkg/k8sops/generators.go]: I think we might want to add a Status subresource here. There's a helpful blog post about it: https://blog.openshift.com/kubernetes-custom-resources-grow-up-in-v1-10/

It only targets 1.10 so we'd have to check the cluster version, but it seems like the idiomatic way to do CRD statuses in 1.10+. Maybe we only want the operator to even target those versions?

**Comment by schallert** [comment]
I left some comments but almost LGTM (I didn't cover the m3admin portion as thoroughly yet). IMO since we have changes planned that will cover a broad portion of the code we can land soon and refactor parts as we touch them rather than holding up feature development.

File: manifests/operator.yaml [hunk 3]
```diff
@@ -72,15 +65,12 @@ spec:
     spec:
```
[the remark this answers is not in the document]
Comment (nerd0) [reply manifests/operator.yaml]: Sure, good call.

File: manifests/operator.yaml [hunk 4]
```diff
@@ -1,10 +1,3 @@
-apiVersion: v1
```
[the remark this answers is not in the document]
Comment (nerd0) [reply manifests/operator.yaml]: All the resource need to be under the namespace though.

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -21,274 +21,202 @@
...
-		if err := c.m3admin.PlacementInit(cluster, placementDetails); err != nil {
-			logrus.WithError(err).Error("failed to init placement, trying again")
-			return false, nil
-		}
-		return true, nil
-	})
-	if err != nil {
-		logrus.WithError(err).Error("failed to init placement after 6 tries")
-		return err
+	if cluster.Spec.ReplicationFactor > int32(len(cluster.Spec.IsolationGroups)) {
```
[the remark this answers is not in the document]
Comment (nerd0) [reply pkg/controller/controller.go]: Sure, fair point

[the remark this answers is not in the document]
Comment (nerd0) [reply pkg/controller/controller.go]: Fair point, updated .

File: manifests/operator.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (nerd0) [reply manifests/operator.yaml]: Added namespace back but can't remove it from the other resources since they need to be under the same namespace.

**Pull request merged**